src/main-oe/sgxlkl_run_oe.c

```diff
@@ -504,30 +440,229 @@ static void sgxlkl_loader_signal_handler(int signo)
+
+        if (!fgets(hash, MAX_HASH_DIGITS + 2, hf))
+            sgxlkl_host_fail(
+                "Failed to read root hash from file %s.\n",
+                verity_file_or_roothash);
+
+        /* Remove possible new line */
+        char* nl = strchr(hash, '\n');
+        if (nl)
+            *nl = 0;
+
+        size_t hash_len = strlen(hash);
+        if (hash_len > MAX_HASH_DIGITS)
+            sgxlkl_host_fail(
+                "Root hash read from file %s too long! Maximum length: %d\n",
+                verity_file_or_roothash,
+                MAX_HASH_DIGITS);
+
+        disk->roothash = (char*)malloc(hash_len + 1);
+        strncpy(disk->roothash, hash, hash_len);
+        disk->roothash[hash_len] = 0;
+
+        fclose(hf);
     }
-    buf[len] = 0;
-    close(fd);
+    else
+        disk->roothash = verity_file_or_roothash;
 
-    if (ret < 0)
+    char* hashoffset_path;
+    if (!verity_file_or_hashoffset)
+    {
+        size_t hashoffset_path_len =
+            strlen(disk_path) + strlen(".hashoffset") + 1;
+        hashoffset_path = (char*)malloc(hashoffset_path_len);
+        snprintf(
+            hashoffset_path,
+            hashoffset_path_len,
+            "%s%s",
+            disk_path,
+            ".hashoffset");
+    }
+    else
+    {
+        hashoffset_path = verity_file_or_hashoffset;
+    }
+
+    char* hashoffset_str;
+    if (access(hashoffset_path, R_OK) != -1)
+    {
+        FILE* hf;
+        char hashoffset_buf[MAX_HASHOFFSET_DIGITS];
+
+        if (!(hf = fopen(hashoffset_path, "r")))
+            sgxlkl_host_fail(
+                "Failed to open hash offset file %s.\n", hashoffset_path);
+
+        if (!fgets(hashoffset_buf, MAX_HASHOFFSET_DIGITS, hf))
+            sgxlkl_host_fail(
+                "Failed to read hash offset from file %s.\n", hashoffset_path);
+
+        fclose(hf);
+
+        hashoffset_str = hashoffset_buf;
+    }
+    else if (verity_file_or_hashoffset)
+    {
+        hashoffset_str = verity_file_or_hashoffset;
+    }
+    else
         sgxlkl_host_fail(
-            "Failed to read %s: %s.\n", app_config_path, strerror(errno));
+            "A hash offset must be set via SGXLKL_HD_VERITY_OFFSET when "
+            "SGXLKL_HD_VERITY is used.\n");
+
+    errno = 0;
+    disk->roothash_offset = strtoll(hashoffset_str, NULL, 10);
+    if (errno == EINVAL || errno == ERANGE)
+        sgxlkl_host_fail("Failed to parse hash offset!\n");
+
+    if (hashoffset_path != verity_file_or_hashoffset)
+        free(hashoffset_path);
+}
+
+static void override_enclave_disk_config(
+    sgxlkl_enclave_root_config_t* disk,
+    const char* image_path,
+    char* keyfile_or_passphrase,
+    char* verity_file_or_roothash,
+    char* verity_file_or_hashoffset)
+{
+    // If key and root hash are provided remotely or is set via enclave
+    // config, don't set it here.
+    if (host_state.enclave_config.mode != HW_RELEASE_MODE)
+    {
+        if (keyfile_or_passphrase)
```

Comment:
I don't think so, it's already it's own function pretty much, and I don't want to break to logic.

Reply:
Hm, seems weird. One is for dm-crypt, the other for dm-verity, and this overall function would just call both smaller functions. Otherwise you might as well inline the verity code into this one.

Reply:
Same difference to me, I just kept the existing structure.